src/Ads/AssetSuggestionsService.php

```diff
@@ -0,0 +1,201 @@
+		$this->wp = $wp;
+	}
+
+	/**
+	 * Get posts that can be used to suggest assets
+	 *
+	 * @param string $search The search query
+	 * @param int    $per_page Number of items per page
+	 *
+	 * @return array
+	 */
+	protected function get_post_suggestions( $search, $per_page ): array {
+		$post_suggestions    = [];
+		$excluded_post_types = [ 'attachment' ];
+
+		$post_types = $this->wp->get_post_types(
+			[
+				'exclude_from_search' => false,
+				'public'              => true,
+			]
+		);
+
+		// Exclude attachment post_type
+		$filtered_post_types = array_diff( $post_types, $excluded_post_types );
+
+		$args = [
+			'post_type'      => $filtered_post_types,
+			'posts_per_page' => $per_page,
+			'post_status'    => 'publish',
+			's'              => $search,
+		];
+
+		$posts = $this->wp->get_posts( $args );
+
+		foreach ( $posts as $post ) {
+			$post_suggestions[] = $this->format_final_url_response( $post->ID, 'post', $post->post_title, get_permalink( $post->ID ) );
+		}
+
+		return $post_suggestions;
+	}
+
+	/**
+	 * Get terms that can be used to suggest assets
+	 *
+	 * @param string $search The search query
+	 * @param int    $per_page Number of items per page
+	 *
+	 * @return array
+	 */
+	protected function get_terms_suggestion( $search, $per_page ): array {
+		$terms_suggestions = [];
+
+		// Get all taxonomies that are public, show_in_menu = true helps to exclude taxonomies such as "product_shipping_class".
+		$taxonomies = $this->wp->get_taxonomies(
+			[
+				'public'       => true,
+				'show_in_menu' => true,
+			],
+		);
+
+		$terms = $this->wp->get_terms(
+			[
+				'taxonomy'   => $taxonomies,
+				'hide_empty' => false,
+				'number'     => $per_page,
+				'name__like' => $search,
+			]
+		);
+
+		foreach ( $terms as $term ) {
+				$terms_suggestions[] = $this->format_final_url_response( $term->term_id, 'term', $term->name, get_term_link( $term->term_id, $term->taxonomy ) );
+		}
+
+		return $terms_suggestions;
+
+	}
+
+
+	/**
+	 * Return a list of final urls that can be used to suggest assets.
+	 *
+	 * @param string $search The search query
+	 * @param int    $per_page Number of items per page
+	 * @param string $order_by Order by: id, type, title, url
+	 *
+	 * @return array Array of final urls with their title, id & type.
+	 */
+	public function get_final_urls_suggestions( $search = '', $per_page = 30, $order_by = 'title' ): array {
+		if ( $per_page <= 0 ) {
+			return [];
+		}
+
+		if ( empty( $search ) ) {
+			return $this->get_defaults_final_urls_suggestions();
+		}
+
+		// Split possible results between posts and terms.
+		$per_page_posts = (int) ceil( $per_page / 2 );
+		$per_page_terms = (int) $per_page - $per_page_posts;
+
+		$posts = $this->get_post_suggestions( $search, $per_page_posts );
+
+		// If not posts results, try to get all results from terms.
+		if ( count( $posts ) === 0 ) {
+			$result = $this->get_terms_suggestion( $search, $per_page );
+			return $this->sort_results( $result, $order_by );
+		}
```

Comment:
As far as I understand this code, if the client enters `$per_page = 10` this code will try to get 5 posts, and 5 terms.

Let's suppose we have 10 terms. In case there are exactly 0 posts, the code will try to get 10 terms to show 10 items. Ok

❓  What if there  is 1-4 posts?  Then the results will be less than 10, even tho, this is not true, since we have 10 terms.

I guess instead of  `if ( count( $posts ) === 0 ) ` it should be `if ( count( $posts ) <  $per_page_posts)` 

However, this piece of code seems unnecessary for me, since we can compute that in the terms call directly.

`$per_page_terms = $per_page - count($posts);`
`$terms  = $this->get_terms_suggestion( $search, $per_page_terms);`

However, we can have still the case in which there is 10 or more posts, and there is less than 5 terms. In that case shouldn't we fill it with the posts?

Reply:
Was thinking about this... Not sure if this strategy of 50%/50%  is something defined as a requirement but maybe it is easier to just return all the posts we can until we complete the "per page" param and fill with terms if necessary (when there are less posts than the per_page param).

Reply:
>However, we can have still the case in which there is 10 or more posts, and there is less than 5 terms. In that case shouldn't we fill it with the posts?

That could be the case, however, every time that the search is more specific, the wanted result will come up if it does not show up in the first request. I tweak a bit the code to try to get more results if we have some margin.

See: https://github.com/woocommerce/google-listings-and-ads/pull/1784/commits/fcf2045afa0bf8076101753c7271fd11c7f8aff9

> it is easier to just return all the posts we can until we complete the "per page" param and fill with terms if necessary (when there are less posts than the per_page param).

The problem with this approach is that if we have X posts matching the query and per_page = X always will return `posts` and the user will not be able to search for terms.